File: polocloud-node/src/main/java/dev/httpmarco/polocloud/node/i18n/I18nProvider.java
package dev.httpmarco.polocloud.node.i18n;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.experimental.Accessors;

import java.util.Collections;
import java.util.Locale;
import java.util.MissingResourceException;
import java.util.ResourceBundle;

@Accessors(fluent = true)
@AllArgsConstructor
public class I18nProvider implements I18n {

    @Getter
    private final String resourceBundlePrefix;
    private final ClassLoader classLoader = getClass().getClassLoader();

    @Override
    public String get(String key) {
        return get(key, Collections.emptyList());
    }

    @Override
    public String get(String key, Object... format) {
        return getDefault(key, format);
    }

    @Override
    public String getDefault(String key, Object... format) {
        return get(key, Locale.ENGLISH, format);
    }

    @Override
    public String get(String key, Locale locale, Object... format) {
        var resourceBundle = resourceBundle(locale);

        if (!resourceBundle.containsKey(key)) {
            return key;
        }

        if (format.length == 0) {
            return resourceBundle.getString(key);
        }

        return String.format(resourceBundle.getString(key), format);
    }

    @Override
    public ResourceBundle resourceBundle(Locale locale) {
        try {
            return ResourceBundle.getBundle(this.resourceBundlePrefix, locale, this.classLoader);
        } catch (MissingResourceException e) {
            return defaultResourceBundle();
        }
    }

    @Override
    public ResourceBundle defaultResourceBundle() {
        return ResourceBundle.getBundle(this.resourceBundlePrefix, Locale.ENGLISH, this.classLoader);
    }
}

File: agent/src/main/kotlin/dev/httpmarco/polocloud/agent/i18n/I18nProvider.kt
package dev.httpmarco.polocloud.agent.i18n

import dev.httpmarco.polocloud.agent.Agent
import dev.httpmarco.polocloud.agent.logger
import dev.httpmarco.polocloud.agent.runtime.local.terminal.LoggingColor
import java.util.*

open class I18nProvider(private val resourceBundlePrefix: String) : I18n {

    private val utf8ResourceBundleControl = UTF8ResourceBundleControl()
    private var overrideLocale: Locale? = null

    val SUPPORTED_LOCALES: List<Locale> = listOf(
        Locale.forLanguageTag("af"),
        Locale.forLanguageTag("ar"),
        Locale.forLanguageTag("ca"),
        Locale.SIMPLIFIED_CHINESE,
        Locale.TRADITIONAL_CHINESE,
        Locale.forLanguageTag("cs"),
        Locale.forLanguageTag("da"),
        Locale.GERMAN,
        Locale.forLanguageTag("el"),
        Locale.ENGLISH,
        Locale.forLanguageTag("es"),
        Locale.forLanguageTag("fi"),
        Locale.FRENCH,
        Locale.forLanguageTag("he"),
        Locale.forLanguageTag("hu"),
        Locale.ITALIAN,
        Locale.JAPANESE,
        Locale.forLanguageTag("ko"),
        Locale.forLanguageTag("nl"),
        Locale.forLanguageTag("no"),
        Locale.forLanguageTag("pl"),
        Locale.forLanguageTag("pt"),
        Locale.forLanguageTag("ro"),
        Locale.forLanguageTag("ru"),
        Locale.forLanguageTag("sr"),
        Locale.forLanguageTag("sv"),
        Locale.forLanguageTag("tr"),
        Locale.forLanguageTag("uk"),
        Locale.forLanguageTag("vi"),
        Locale.forLanguageTag("zh")
    )

    fun info(key: String, vararg format: Any?) {
        logger.info(get(key, *format))
    }

    fun warn(key: String, vararg format: Any?) {
        logger.warn(get(key, *format))
    }

    fun error(key: String, vararg format: Any?) {
        logger.error(get(key, *format))
    }

    fun debug(key: String, vararg format: Any?) {
        logger.debug(get(key, *format))
    }

    override fun get(key: String): String {
        return get(key, *emptyArray())
    }

    override fun get(key: String, vararg format: Any?): String {
        val locale = try {
            overrideLocale ?: Agent.instance.config.locale
        } catch (ex: Throwable) {
            return getDefault(key, *format) // Fallback for startup
        }

        return get(key, locale, *format)
    }

    override fun get(key: String, locale: Locale, vararg format: Any?): String {
        return translate(resourceBundle(locale), key, *format)
    }

    override fun getDefault(key: String, vararg format: Any?): String {
        return get(key, Locale.ENGLISH, *format)
    }

    override fun resourceBundle(locale: Locale): ResourceBundle {
        return try {
            this.localBundle(locale)
        } catch (_: MissingResourceException) {
            defaultResourceBundle()
        }
    }

    override fun defaultResourceBundle(): ResourceBundle {
        return this.localBundle(Locale.ENGLISH)
    }

    private fun localBundle(locale: Locale): ResourceBundle {
        return ResourceBundle.getBundle(
            this.resourceBundlePrefix,
            locale,
            this.javaClass.classLoader,
            utf8ResourceBundleControl
        )
    }

    private fun translate(resourceBundle: ResourceBundle, key: String, vararg format: Any?): String {
        if (!resourceBundle.containsKey(key)) {
            return key
        }

        val value = resourceBundle.getString(key)
        return if (format.isEmpty()) {
            LoggingColor.translate(value)
        } else {
            LoggingColor.translate(String.format(value, *format))
        }
    }

    fun overrideLocale(locale: Locale) {
        this.overrideLocale = locale
    }

    fun clearOverrideLocale() {
        this.overrideLocale = null
    }
}